File: 第一阶段/1.2-变量与类型/variables.py
# Python 变量与类型示例
# 对应文章：1.2 变量与类型系统

print("=== 基础赋值 ===")
# Python 天生就是"推断"的，不需要声明类型
x = 5              # int
name = "张三"      # str
price = 9.99       # float
is_true = True     # bool

# 也可以加类型注解（但解释器不理你，只给 IDE 和 mypy 用）
y: int = 10
name2: str = "李四"
print(f"x={x}, y={y}, name2={name2}")

# ==================== 变量可以改变类型（动态类型！） ====================
print("\n=== 动态类型 ===")
x = 5
print(f"x = {x}, type = {type(x)}")       # <class 'int'>
x = "hello"
print(f"x = {x}, type = {type(x)}")       # <class 'str'>
# C# 对比：C# 的 var 一旦推断出类型就不能变了

# ==================== 类型检查 ====================
print("\n=== 类型检查 ===")
print(f"isinstance(5, int)    = {isinstance(5, int)}")       # True
print(f"isinstance(5, (int, float)) = {isinstance(5, (int, float))}")  # True（可以检查多个类型）
print(f"type(5).__name__      = {type(5).__name__}")          # "int"
# 推荐用 isinstance() 而不是 type() == ，因为 isinstance 支持继承

# ==================== 类型转换 ====================
print("\n=== 类型转换 ===")
s = "123"
n = int(s)                     # 123
f = float("3.14")              # 3.14
b = bool(0)                    # False
b2 = bool(1)                   # True
b3 = bool("")                  # False
print(f"int('123')      = {n}")
print(f"float('3.14')   = {f}")
print(f"bool(0)         = {b}")
print(f"bool('')        = {b3}")

# 安全转换（Python 没有 TryParse，需要 try-except）
def safe_int(s):
    """安全地将字符串转为整数，失败返回 (None, False)"""
    try:
        return int(s), True
    except ValueError:
        return None, False

val, ok = safe_int("abc")
print(f"safe_int('abc') = ({val}, {ok})")
val, ok = safe_int("42")
print(f"safe_int('42')  = ({val}, {ok})")

# ==================== 常量约定 ====================
print("\n=== 常量约定 ===")
# Python 没有 const 关键字，靠全大写命名约定
PI = 3.14159
GREETING = "Hello"
# PI = 999  # 不会报错！但违反约定，linter 会警告
print(f"PI = {PI}")

# 如果需要真正不可变，可以用 dataclass(frozen=True) 或 typing.Final
from typing import Final
MAX_SIZE: Final = 100  # 类型提示：这是一个常量
print(f"MAX_SIZE = {MAX_SIZE}")

File: 第一阶段/1.2-变量与类型/Variables.cs
// C# 变量与类型示例
// 对应文章：1.2 变量与类型系统

using System;
using System.Collections.Generic;

// ==================== 显式类型声明 ====================
Console.WriteLine("=== 显式类型声明 ===");
int x = 5;
string name = "张三";
double price = 9.99;
bool isTrue = true;

// ==================== 类型推断（var）====================
Console.WriteLine("\n=== 类型推断 var ===");
var y = 10;        // 编译器推断为 int（一旦确定不能变）
var msg = "hello"; // 编译器推断为 string
Console.WriteLine($"y = {y}, msg = {msg}");

// ==================== 动态类型 ====================
Console.WriteLine("\n=== 动态类型 dynamic ===");
dynamic z = 5;     // 运行时才确定类型
z = "hello";       // 可以改变类型（但失去编译时类型安全！）
Console.WriteLine($"z = {z}");

// ==================== 类型检查 ====================
Console.WriteLine("\n=== 类型检查 ===");
Console.WriteLine($"x is int:       {x is int}");           // true
Console.WriteLine($"x.GetType():    {x.GetType()}");        // System.Int32
Console.WriteLine($"x is > 0:       {x is > 0}");           // true（模式匹配）

// ==================== 类型转换 ====================
Console.WriteLine("\n=== 类型转换 ===");
string s = "123";
int n = int.Parse(s);                          // 123（失败抛异常）
bool success = int.TryParse(s, out int result);  // 安全转换（推荐！）
Console.WriteLine($"int.Parse('123')   = {n}");
Console.WriteLine($"TryParse('123')    = {success}, {result}");
Console.WriteLine($"TryParse('abc')    = {int.TryParse("abc", out _)}, {int.TryParse("abc", out int r2)}");

// ==================== 常量 ====================
Console.WriteLine("\n=== 常量 ===");
const double Pi = 3.14159;     // 编译时常量
const string Greeting = "Hello";
// Pi = 999;  // 编译错误！常量不可修改
Console.WriteLine($"Pi = {Pi}");

// readonly（运行时常量，可在构造函数中赋值）
readonly int MaxRetries = 3;   // 只能在类中使用
